File: firebase-auth/src/main/java/com/androidhuman/firebase/auth/RxFirebaseUser.java
package com.androidhuman.firebase.auth;

import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.UserProfileChangeRequest;

import com.androidhuman.firebase.auth.model.TaskResult;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;

import rx.Observable;

public final class RxFirebaseUser {

    @CheckResult
    @NonNull
    public static Observable<TaskResult> delete(@NonNull FirebaseUser user) {
        return Observable.create(new UserDeleteOnSubscribe(user));
    }

    @CheckResult
    @NonNull
    public static Observable<String> getToken(
            @NonNull FirebaseUser user, boolean forceRefresh) {
        return Observable.create(new UserGetTokenOnSubscribe(user, forceRefresh));
    }

    @CheckResult
    @NonNull
    public static Observable<AuthResult> linkWithCredential(
            @NonNull FirebaseUser user, @NonNull AuthCredential credential) {
        return Observable.create(new UserLinkWithCredentialOnSubscribe(user, credential));
    }

    @CheckResult
    @NonNull
    public static Observable<TaskResult> reauthenticate(
            @NonNull FirebaseUser user, @NonNull AuthCredential credential) {
        return Observable.create(new UserReauthenticateOnSubscribe(user, credential));
    }

    @CheckResult
    @NonNull
    public static Observable<TaskResult> reload(@NonNull FirebaseUser user) {
        return Observable.create(new UserReloadOnSubscribe(user));
    }

    @CheckResult
    @NonNull
    public static Observable<TaskResult> sendEmailVerification(@NonNull FirebaseUser user) {
        return Observable.create(new UserSendEmailVerificationOnSubscribe(user));
    }

    @CheckResult
    @NonNull
    public static Observable<AuthResult> unlink(
            @NonNull FirebaseUser user, @NonNull String provider) {
        return Observable.create(new UserUnlinkOnSubscribe(user, provider));
    }

    @CheckResult
    @NonNull
    public static Observable<TaskResult> updateEmail(
            @NonNull FirebaseUser user, @NonNull String email) {
        return Observable.create(new UserUpdateEmailOnSubscribe(user, email));
    }

    @CheckResult
    @NonNull
    public static Observable<TaskResult> updatePassword(
            @NonNull FirebaseUser user, @NonNull String password) {
        return Observable.create(new UserUpdatePasswordOnSubscribe(user, password));
    }

    @CheckResult
    @NonNull
    public static Observable<TaskResult> updateProfile(
            @NonNull FirebaseUser user, @NonNull UserProfileChangeRequest request) {
        return Observable.create(new UserUpdateProfileOnSubscribe(user, request));
    }

    private RxFirebaseUser() {
        throw new AssertionError("No instances");
    }
}

File: firebase-auth-kotlin/src/main/kotlin/com/androidhuman/firebase/auth/RxFirebaseUser.kt
@file:Suppress("NOTHING_TO_INLINE")

package com.androidhuman.firebase.auth

import com.androidhuman.firebase.auth.model.TaskResult
import com.google.firebase.auth.AuthCredential
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.UserProfileChangeRequest
import rx.Observable

inline fun FirebaseUser.rxDelete()
        : Observable<TaskResult>
        = RxFirebaseUser.delete(this)

inline fun FirebaseUser.rxGetToken(forceRefresh: Boolean)
        : Observable<String>
        = RxFirebaseUser.getToken(this, forceRefresh)

inline fun FirebaseUser.rxLinkWithCredential(credential: AuthCredential)
        : Observable<AuthResult>
        = RxFirebaseUser.linkWithCredential(this, credential)

inline fun FirebaseUser.rxReauthenticate(credential: AuthCredential)
        : Observable<TaskResult>
        = RxFirebaseUser.reauthenticate(this, credential)

inline fun FirebaseUser.rxReload()
        : Observable<TaskResult>
        = RxFirebaseUser.reload(this)

inline fun FirebaseUser.rxSendEmailVerification()
        : Observable<TaskResult>
        = RxFirebaseUser.sendEmailVerification(this)

inline fun FirebaseUser.rxUnlink(provider: String)
        : Observable<AuthResult>
        = RxFirebaseUser.unlink(this, provider)

inline fun FirebaseUser.rxUpdateEmail(email: String)
        : Observable<TaskResult>
        = RxFirebaseUser.updateEmail(this, email)

inline fun FirebaseUser.rxUpdatePassword(password: String)
        : Observable<TaskResult>
        = RxFirebaseUser.updatePassword(this, password)

inline fun FirebaseUser.rxUpdateProfile(request: UserProfileChangeRequest)
        : Observable<TaskResult>
        = RxFirebaseUser.updateProfile(this, request)

